iocage_lib/ioc_common.py

```diff
@@ -654,72 +654,68 @@ def check_release_newer(release, callback=None, silent=False):
-        for example 'foo' will be contructed into [foo=IOCAGE_GEN_RULENUM]
-
-    Will construct a devfs ruleset from dict(paths), and list(includes) with:
-        paths dict:
-            EXAMPLE: "usbctl": mode 660 group uucp
-            - path to unhide: mode (can be None)
-        includes list:
-            EXAMPLE ["$devfsrules_hide_all", "$devfsrules_unhide_basic"]
-            - [entry, entry]
-
-    Returns a tuple containing the string and which devfs_ruleset got assigned.
-    """
-    includes = [] if includes is None else includes
-    ct_str = f'## {ruleset_name}' if comment is None else comment
-    rules = []
-    ruleset_number = 5  # The system has 4 already claimed.
-
-    try:
-        with open('/etc/devfs.rules', 'r') as f:
-            exists = False
+    Will add a per jail devfs ruleset with the specified rules
 
-            for line in f.readlines():
-                if line.rstrip() == ct_str:
-                    exists = True
-                    continue
-
-                if exists:
-                    return None, int(line.rsplit('=')[1].strip(']\n'))
+    Returns the devfs_ruleset number that has been allocated.
+    """
+    ruleset = 5  # 0-4 is always reserved
+    devfs_rulesets = su.run(
+        ['devfs', 'rule', 'showsets'],
+        stdout=su.PIPE, universal_newlines=True
+    )
+    ruleset_list = [int(i) for i in devfs_rulesets.stdout.splitlines()]
+    devfs_dict = {
+        'zfs': None
+    }
+    devfs_includes = [
+        '1',  # '$devfsrules_hide_all'
+        '2',  # '$devfsrules_unhide_basic'
+        '3',  # '$devfsrules_unhide_login'
+    ]
 
-                if line.startswith('['):
-                    try:
-                        line = int(line.rsplit('=')[1].strip(']\n'))
-                    except IndexError:
-                        rules.append(line)
-                    rules.append(line)
-    except FileNotFoundError:
-        logit(
-            {
-                'level': 'EXCEPTION',
-                'message':
-                    '/etc/devfs.rules could not be found, unable to continue.'
-            }
+    while ruleset in ruleset_list:
+        ruleset += 1
+
+    ruleset = str(ruleset)
+    if paths is not None:
+        devfs_dict.update(paths)
+
+    if includes is not None:
+        devfs_includes += includes
+
+    # We may end up setting all of these.
+    if conf['allow_tun'] == '1':
+        devfs_dict['tun*'] = None
+    if conf['dhcp'] == 'on':
+        devfs_dict['bpf*'] = None
+
+    for include in devfs_includes:
+        # In case a plugin or something tries to use the human readable names
+        if include == '$devfsrules_hide_all':
+            include = '1'
+        elif include == '$devfsrules_unhide_basic':
+            include = '2'
+        elif include == '$devfsrules_unhide_login':
+            include = '3'
+
+        su.run(
+            ['devfs', 'rule', '-s', ruleset, 'add', 'include', include],
```

Comment:
Hmm, why do we need to operate on /dev, and not on the devfs insided the jail?

Reply:
While we could operate on the jails /dev, users are able to specify their own rulesets. In which case those won't exist on the jail's /dev, and we can also get the devfs rules set up earlier then starting the jail by using the hosts. I don't see an advantage to using the jail's /dev.